### PR TITLE
ASoC: SOF: ipc4: pause/reset for pipeline without AIF

### PR DESCRIPTION
**Pull request by libinyang**
Some pipelines don't have an AIF widget. In this case, the pipelines won't have chance to pause/reset. This will cause the pipelines only work at the first time. After stopping/pausing the stream, the pipelines won't work any more.

This patch allows the pause/reset operation on the pipelines which don't have an AIF widget.

Signed-off-by: Libin Yang <libin.yang@intel.com>

### CODE REVIEW

**Comment by libinyang** [comment]
If the tplg pipeline does't have an AIF widget, like the pipeline 14 in the attached picture, the pipeline don't have chance to pause or reset when the stream is paused or reset. This causes resuming / reopening the stream fails. This patch tries to fix the pipeline without an AIF widget pause/reset issue.

cavs-tgl-nocodec

**Comment by plbossart** [comment]
@libinyang the pipeline 14 in the picture above is similar to the others, it starts with a DAI. Not sure what the differences are and why we need to special-case it. Please explain a bit more the root-cause and the problem statement.

File: sound/soc/sof/ipc4-pcm.c [hunk 1]
```diff
@@ -65,7 +65,7 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 		switch (state) {
 		case SOF_IPC4_PIPE_PAUSED:
 		case SOF_IPC4_PIPE_RESET:
-			if (!WIDGET_IS_AIF(swidget->id))
+			if (WIDGET_IS_DAI(swidget->id))
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-pcm.c]: The original `if (!WIDGET_IS_AIF(swidget->id))` _included_ the new `if (WIDGET_IS_DAI(swidget->id))`, but now the code will also skip AIF for PAUSE/RESET which is against what the comment was saying that the PAUSE/RESET is *only* applied to FE (AIF).

Are you sure that this is correct?

Comment (ranj063) [reply sound/soc/sof/ipc4-pcm.c]: @libinyang the intention for the !(WIDGET_IS_AIF()) check is to make sure that only the FE pipelines are triggered during the FE CPU DAI trigger. So changing this check is wrong. But I understand where you are coming from. In the case of IPC3, we used the widget's DAPM POWER_UP/DOWN event to trigger the detect pipeline (pipeline 15). So you have 2 choices here: 1. Stick with the same solution as IPC3 2. Expand the check here to also trigger the pipelines associated with widgets that have the same stream name as the host pipeline. I prefer that latter as that will be generic enough to allow other connected pipelines to get triggered as well

**Comment by libinyang** [comment]
If the tplg pipeline does't have an AIF widget, like the pipeline 14 in the attached picture, the pipeline don't have chance to pause or reset when the stream is paused or reset. This causes resuming / reopening the stream fails. This patch tries to fix the pipeline without an AIF widget pause/reset issue.

cavs-tgl-nocodec


> @libinyang the pipeline 14 in the picture above is similar to the others, it starts with a DAI. Not sure what the differences are and why we need to special-case it. Please explain a bit more the root-cause and the problem statement.

@plbossart In IPC4 mode, when sof_ipc4_trigger_pipelines() is called to set the pipeline to pause or reset state, it will find the pipeline's AIF widget to do the setting, which is different from IPC3. If the pipeline don't have AIF widget in IPC4, the pipeline will have no chance to change to pause or reset state. In the previous case, as you can see, there are 3 pipelines in the DMIC WoV stream. Pipeline 13  contains copier.DMIC.14.1 and kpb.14.1; pipeline 14 contains micsel.15.1, intelwov.15.1 and virtual.detect_sink; pipeline 13 contains copier.host.13.1. Pipeline 13 and pipeline 14 are both OK to change state to pause/reset because they contains either AIF widget or DAI widget (pipeline with DAI widget will set state in other function). But pipeline 15 has problem. sof_ipc4_trigger_pipelines() finds the pipeline doesn't contain AIF widget, so this pipeline will never have change to set to pause/reset state.

File: sound/soc/sof/ipc4-pcm.c (hunk 1 above)
Comment (libinyang) [reply sound/soc/sof/ipc4-pcm.c]: > The original `if (!WIDGET_IS_AIF(swidget->id))` _included_ the new `if (WIDGET_IS_DAI(swidget->id))`, but now the code will also skip AIF for PAUSE/RESET which is against what the comment was saying that the PAUSE/RESET is _only_ applied to FE (AIF).
> 
> Are you sure that this is correct?

@ujfalusi When pipeline is set to pause/reset state, the pipeline must contain AIF widget. The patch allow pipeline to change to pause/reset state even it doesn't contain AIF widget. PS: Please see the special case why we need this patch in the comment https://github.com/thesofproject/linux/pull/3866#issuecomment-1251728904.

Comment (libinyang) [reply sound/soc/sof/ipc4-pcm.c]: > @libinyang the intention for the !(WIDGET_IS_AIF()) check is to make sure that only the FE pipelines are triggered during the FE CPU DAI trigger. So changing this check is wrong. But I understand where you are coming from. In the case of IPC3, we used the widget's DAPM POWER_UP/DOWN event to trigger the detect pipeline (pipeline 15). So you have 2 choices here: 1. Stick with the same solution as IPC3 2. Expand the check here to also trigger the pipelines associated with widgets that have the same stream name as the host pipeline. I prefer that latter as that will be generic enough to allow other connected pipelines to get triggered as well

@ranj063 Do you mean we should not pause/reset the pipeline 14 in our case in the function sof_ipc4_trigger_pipelines()?  Is there any reason to prevent this? For the second solution you recommended, this requires the widget virtual.detect_sink has the same name as copier.host.13.1. But if there is a case that the widgets like virtual.detect_sink is connected to several widgets, then it's hard to set the stream name of it.

Comment (ranj063) [reply sound/soc/sof/ipc4-pcm.c]: Why do we need the same name? Why not set the stream name for all widgets in pipeline 15 to be copier.host.13. And trigger it too because pipeline 13 was triggered

Comment (libinyang) [reply sound/soc/sof/ipc4-pcm.c]: @ranj063 I mean if pipeline 15 is also connected to another piepline, let's say pipeline 16 copier.host.16.1, we can't set the pipeline 15 widgets stream name the same as both pipeline 13 and pipeline 16 as ppl13 and ppl16 name may be different.

**Comment by plbossart** [comment]
> If the tplg pipeline does't have an AIF widget, like the pipeline 14 in the attached picture, the pipeline don't have chance to pause or reset when the stream is paused or reset. This causes resuming / reopening the stream fails. This patch tries to fix the pipeline without an AIF widget pause/reset issue.
> 
> cavs-tgl-nocodec
> 
> > @libinyang the pipeline 14 in the picture above is similar to the others, it starts with a DAI. Not sure what the differences are and why we need to special-case it. Please explain a bit more the root-cause and the problem statement.
> 
> @plbossart In IPC4 mode, when sof_ipc4_trigger_pipelines() is called to set the pipeline to pause or reset state, it will find the pipeline's AIF widget to do the setting, which is different from IPC3. If the pipeline don't have AIF widget in IPC4, the pipeline will have no chance to change to pause or reset state. In the previous case, as you can see, there are 3 pipelines in the DMIC WoV stream. Pipeline 13 contains copier.DMIC.14.1 and kpb.14.1; pipeline 14 contains micsel.15.1, intelwov.15.1 and virtual.detect_sink; pipeline 13 contains copier.host.13.1. Pipeline 13 and pipeline 14 are both OK to change state to pause/reset because they contains either AIF widget or DAI widget (pipeline with DAI widget will set state in other function). But pipeline 15 has problem. sof_ipc4_trigger_pipelines() finds the pipeline doesn't contain AIF widget, so this pipeline will never have change to set to pause/reset state.

Sorry @libinyang, you are mixing pipeline 14 and 15 in your explanations and making it really hard to follow your line of though. I don't see how a pipeline 14 could contains a micsel.15.1, that's really wrong even if it was true.

Based on the image only, it looks like only pipeline 15 is not connected to a DAI or host. I don't see how pipeline 14 is problematic.

File: sound/soc/sof/ipc4-pcm.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-pcm.c]: @libinyang my point is we need a generic way to handle intermediate pipelines. host pipelines will be triggered in the FE CPU DAI trigger and DAI pipelines will be triggered in the BE CPU DAI trigger. So the question is what happens to the pipelines in between? And that should be determined by which pipeline they should be trigger alongside with ie host/dai pipeline and this can be indicated with the stream name for the widgets in the intermediate pipelines. So in your example, if there are 2 intermediate pipelines 15 and 16, pipeline 15 would have the stream name copier.host.13 to say that it should be triggered in the FE CPU DAI trigger and pipeline 16 (hypothetically speaking) will have the stream name copier.DMIC.14.1 to say that it should be trigger in the BE CPU DAi trigger

**Comment by libinyang** [comment]
> Sorry @libinyang, you are mixing pipeline 14 and 15 in your explanations and making it really hard to follow your line of though. I don't see how a pipeline 14 could contains a micsel.15.1, that's really wrong even if it was true.

@plbossart Sorry. It should be "Pipeline 14 contains copier.DMIC.14.1 and kpb.14.1; pipeline 15 contains micsel.15.1, intelwov.15.1 and virtual.detect_sink; pipeline 13 contains copier.host.13.1. "

> 
> Based on the image only, it looks like only pipeline 15 is not connected to a DAI or host. I don't see how pipeline 14 is problematic.

Yes. Pipeline 14 OK. The problem is pipeline 15. Pipeline 15 won't be triggered to pause/reset state as it is not connected to a DAI or host widget.

File: sound/soc/sof/ipc4-pcm.c (hunk 1 above)
Comment (libinyang) [reply sound/soc/sof/ipc4-pcm.c]: @ranj063 I mean the below topology:
```
ppl 1 --- ppl2 --- ppl3 (host)    
            |-------ppl4 (host)
```
How can we name the ppl2 widget's stream name, name after ppl3 or ppl4?

And for the solution 1 of using POWER_UP/DOWN event, I still have some concerns:
1. In this case, every intermediate ppl needs such event. It's a little complicated. 
2. intermediate ppl will be triggered start twice. In the POWER_UP event, it will be triggered. And in the sof_ipc4_trigger_pipelines(), it will still be triggered start as we allow all the widgets to be triggered start in this function.
3. We have requirement of the trigger order of the pipelines as Kai mentioned. But if we are using POWER_UP/DOWN event, the trigger order may be wrong.

What do you think if we have a flag in tplg, telling it's a widget on DAI pipline, and we will skip triggering them in the function sof_ipc4_trigger_pipelines()?

**Comment by ujfalusi** [comment]
image
@libinyang, what are the dotted lines?
@plbossart, both pipeline 14 and pipeline 15 connected to DAI via `kpb.14.1`
The current code would only place the pipeline 13 to RESET/PAUSED state, which is arguably correct or not.

I guess the question boils down to this: if the pipeline14 -> pipeline13 capture is paused/stopped, do we want to keep the WoV running in the background or they need to be stopped as well? 
Why those widgets can be part of the pipeline13?

**Comment by libinyang** [comment]
> @libinyang, what are the dotted lines? 

Dotted widget is a virtual widget and dotted lines means virtual connection. It doesn't exist in the firmware.

>@plbossart, both pipeline 14 and pipeline 15 connected to DAI via `kpb.14.1` The current code would only place the pipeline 13 to RESET/PAUSED state, which is arguably correct or not.
> 
> I guess the question boils down to this: if the pipeline14 -> pipeline13 capture is paused/stopped, do we want to keep the WoV running in the background or they need to be stopped as well? Why those widgets can be part of the pipeline13?

We must pause/reset ppl14 in this case. Otherwise, the second running capture will fail. Moving the widgets to ppl13 is a workaround, and it is not a root fix.

**Comment by ujfalusi** [comment]
> > @libinyang, what are the dotted lines?
> 
> Dotted widget is a virtual widget and dotted lines means virtual connection. It doesn't exist in the firmware.

OK, so it is marked as virtual in tplg and no IPC is sent for them?

> > @plbossart, both pipeline 14 and pipeline 15 connected to DAI via `kpb.14.1` The current code would only place the pipeline 13 to RESET/PAUSED state, which is arguably correct or not.
> > I guess the question boils down to this: if the pipeline14 -> pipeline13 capture is paused/stopped, do we want to keep the WoV running in the background or they need to be stopped as well? Why those widgets can be part of the pipeline13?
> 
> We must pause/reset ppl14 in this case. Otherwise, the second running capture will fail. Moving the widgets to ppl13 is a workaround, and it is not a root fix.

Are you saying that the second capture is always broken? Even without pipeline 15 in the picture? Just a straight pipeline 13 -> pipeline 14 will not work on the second try?

**Comment by plbossart** [comment]
@ranj063 @libinyang @ujfalusi I feel that we should come up with a simpler test where we chain three pipelines artificially and check how the pipeline states are handled. We also have real cases such as the two-stage mixer where we do have an intermediate pipeline, see what I added for topology1 in https://github.com/thesofproject/sof/pull/6131
sof-adl-nocodec
Pipeline 13 in this topology is not connected to host or dai as well, this is likely to be broken with IPC4. 

Doing this with a complicated WoV setup is more difficult and challenging IMHO.

**Comment by libinyang** [comment]
> Are you saying that the second capture is always broken? Even without pipeline 15 in the picture? Just a straight pipeline 13 -> pipeline 14 will not work on the second try?

ppl14 has a DAI widget and it will be set in another function. So it is OK. Ppl15 has neither DAI widget nor AIF widget and this doesn't work.

**Comment by ujfalusi** [comment]
> > Are you saying that the second capture is always broken? Even without pipeline 15 in the picture? Just a straight pipeline 13 -> pipeline 14 will not work on the second try?
> 
> ppl14 has a DAI widget and it will be set in another function. So it is OK. Ppl15 has neither DAI widget nor AIF widget and this doesn't work.

people 15 is dot connected to 'virtual.detect_sink' and that is dot connected to AIF on copier..host.13.1.

Pipeline 15 got enabled at some point (I guess) since it has a connection from DAI to some sink (AIF or loopback).

I guess the issue can also be present when the path contains more than **2** pipelines, one connected to DAI, one connecte to AIF and X in-between these and only the two end is going to be RESET/PAUSE.

In a way along with the FE we need to handle everything which does not contain DAI, right?
Or on the BE handling we should do the same?

In any case this needs to be clearly documented at minimum.

**Comment by ranj063** [comment]
> @ranj063 @libinyang @ujfalusi I feel that we should come up with a simpler test where we chain three pipelines artificially and check how the pipeline states are handled. We also have real cases such as the two-stage mixer where we do have an intermediate pipeline, see what I added for topology1 in [thesofproject/sof#6131](https://github.com/thesofproject/sof/pull/6131)

@plbossart this is a good idea. I think what we need to do in the driver is to do something very similar to what we used for walking the list of widgets from source->sink while setting them up ie start the pipelines starting from the source and ending at the last pipeline that is the not BE DAI pipeline in the FE CPU DAI trigger. This way we can honor the order of triggering the pipelines. 

Then the other open will be how to tell if a pipeline is the BE DAI pipeline or not. Maybe we can add a token for the pipeline widget qualifying it as a BE DAI pipeline in topology. 

Lastly, as for the wov topology where the detect pipeline is parallel to the host pipeline, the order shouldnt matter. @libinyang please correct me if I am wrong

**Comment by libinyang** [comment]
> @plbossart this is a good idea. I think what we need to do in the driver is to do something very similar to what we used for walking the list of widgets from source->sink while setting them up ie start the pipelines starting from the source and ending at the last pipeline that is the not BE DAI pipeline in the FE CPU DAI trigger. This way we can honor the order of triggering the pipelines.
> 
> Then the other open will be how to tell if a pipeline is the BE DAI pipeline or not. Maybe we can add a token for the pipeline widget qualifying it as a BE DAI pipeline in topology.

Agree. I think the token is a simple solution for us.

> 
> Lastly, as for the wov topology where the detect pipeline is parallel to the host pipeline, the order shouldnt matter. @libinyang please correct me if I am wrong

Yes, it should work. Another solution is we can use multiple pipeline set state IPC message. IPC4 supports set multiple pipeline state at the same time. 
The code prototype, which can be added in the sof_ipc4_trigger_pipelines(), is like:
```
/*
 * pipeline cnt is less then 256, each bit of pipelines[] presents a pipeline. 
 * When the bit is set to 1, it means the corresponding ppl needs to set state 
 * /
u32 pipelines[8]; 
for_each_dapm_widgets(list, num_widgets, widget) {
    swidget = widget->dobj.private;
    if (pipeline_need_set_state(swidget) {
        ppl_id = swidget->pipeline_id;
        set_ppl_bit(pipelines, ppl_id);
    }
    multiple_pipeline_set_state(sdev, pipelines, state);
}
```
In the multiple_pipeline_set_state(), we just need to setup the primary, extension and the payload based on the pipelines[] value and send the IPC message.

**Comment by libinyang** [comment]
> people 15 is dot connected to 'virtual.detect_sink' and that is dot connected to AIF on copier..host.13.1.
> 
> Pipeline 15 got enabled at some point (I guess) since it has a connection from DAI to some sink (AIF or loopback).
> 
> I guess the issue can also be present when the path contains more than **2** pipelines, one connected to DAI, one connecte to AIF and X in-between these and only the two end is going to be RESET/PAUSE.

Yes.

> 
> In a way along with the FE we need to handle everything which does not contain DAI, right? Or on the BE handling we should do the same?
> 
> In any case this needs to be clearly documented at minimum.

My understand is we have handled the DAI pipeline correctly somewhere else. This is why we don't handle DAI pipeline in sof_ipc4_trigger_pipelines(), to avoid the DAI pipeline is paused/reset twice. @ranj063 Please correct me if I'm wrong.

**Comment by libinyang** [comment]
> @ranj063 @libinyang @ujfalusi I feel that we should come up with a simpler test where we chain three pipelines artificially and check how the pipeline states are handled. We also have real cases such as the two-stage mixer where we do have an intermediate pipeline, see what I added for topology1 in [thesofproject/sof#6131](https://github.com/thesofproject/sof/pull/6131) 

@plbossart I haven't tested your tplg because I don't have the platform and the tplg is for IPC3, not for IPC4. So I created a tplg for tglh-nocodec. With the patch the playback can run several times, while without the patch the playback fails at the second time. Please see the attachment for the log.
cavs-tgl-nocodec

[dmesg_playback_without_patch.log](https://github.com/thesofproject/linux/files/9652951/dmesg_playback_without_patch.log)
[dmesg_playback_with_patch.log](https://github.com/thesofproject/linux/files/9652953/dmesg_playback_with_patch.log)
[playback_without_patch.log](https://github.com/thesofproject/linux/files/9652957/playback_without_patch.log)
[playback_with_patch.log](https://github.com/thesofproject/linux/files/9652958/playback_with_patch.log)

@plbossart @ranj063 @ujfalusi I will refine the patch to use a token in the tplg to address whether the widget is on DAI pipeline or not. For the trigger order issue, this is another topic and the solution is not determined. Let's open a new PR topic for it.

**Comment by libinyang** [comment]
v2:
- add the token on_be_ppl support. Driver will check whether the widget is on BE pipeline or not when it handles pause/reset message in sof_ipc4_trigger_pipelines()

**Comment by libinyang** [comment]
> @plbossart @ranj063 @ujfalusi I will refine the patch to use a token in the tplg to address whether the widget is on DAI pipeline or not. For the trigger order issue, this is another topic and the solution is not determined. Let's open a new PR topic for it.

I saw Rander was submitting the patch to add the gain support to cavs-nocodec. His patch doesn't have intermediate pipeline. I will submit another patch based on Rander's patch to separate the stream into 3 pipelines. After that, I think we can merge this PR together with the topology patch.

File: include/uapi/sound/sof/tokens.h [hunk 2]
```diff
@@ -96,6 +96,7 @@
  */
 #define SOF_TKN_COMP_SINK_PIN_BINDING_WNAME	413
 #define SOF_TKN_COMP_SRC_PIN_BINDING_WNAME	414
+#define SOF_TKN_COMP_ON_BE_PIPELINE		416
```
Comment (ranj063) [inline include/uapi/sound/sof/tokens.h:99]: @libinyang this PR only partially addresses your specific problem for triggering the detect pipeline for WoV. It does not address the generic case of triggering multiple pipelines when there are one or more intermediate pipelines. In such cases, we also need to ensure the order in which the pipelines are triggered. Walking the widgets in the DAPM list, will not guarantee the ordering required by the FW

Comment (libinyang) [reply include/uapi/sound/sof/tokens.h:99]: > @libinyang this PR only partially addresses your specific problem for triggering the detect pipeline for WoV. It does not address the generic case of triggering multiple pipelines when there are one or more intermediate pipelines. In such cases, we also need to ensure the order in which the pipelines are triggered. Walking the widgets in the DAPM list, will not guarantee the ordering required by the FW

@ranj063 Yes, this patch won't fix the trigger order issue. The trigger order is another issue. I'm planning to create another PR for it.

**Comment by ranj063** [comment]
@libinyang I have a replacement PR for this one: https://github.com/thesofproject/linux/pull/3914. Basically you only need to set the token for the pipeline and not all the widgets in the BE pipeline.

**Comment by libinyang** [comment]
Close this PR as Ranjani has a new PR for it.
